apps/heft/src/plugins/CopyGlobsPlugin.ts

```diff
@@ -0,0 +1,173 @@
+};
+
+const MAX_PARALLELISM: number = 100;
+
+export class CopyGlobsPlugin implements IHeftPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    const logger: ScopedLogger = heftSession.requestScopedLogger('copy-globs');
+    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
+      build.hooks.preCompile.tap(PLUGIN_NAME, (preCompile: IPreCompileSubstage) => {
+        preCompile.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runCopyFilesForHeftEvent(HeftEvent.preCompile, logger, heftConfiguration);
+        });
+      });
+
+      build.hooks.compile.tap(PLUGIN_NAME, (compile: ICompileSubstage) => {
+        compile.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runCopyFilesForHeftEvent(HeftEvent.compile, logger, heftConfiguration);
+        });
+      });
+
+      build.hooks.bundle.tap(PLUGIN_NAME, (bundle: IBundleSubstage) => {
+        bundle.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runCopyFilesForHeftEvent(HeftEvent.bundle, logger, heftConfiguration);
+        });
+      });
+
+      build.hooks.postBuild.tap(PLUGIN_NAME, (postBuild: IPostBuildSubstage) => {
+        postBuild.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runCopyFilesForHeftEvent(HeftEvent.postBuild, logger, heftConfiguration);
+        });
+      });
+    });
+  }
+
+  private async _runCopyFilesForHeftEvent(
+    heftEvent: HeftEvent,
+    logger: ScopedLogger,
+    heftConfiguration: HeftConfiguration
+  ): Promise<void> {
+    const eventActions: IHeftEventActions = await CoreConfigFiles.getConfigConfigFileEventActionsAsync(
+      logger.terminal,
+      heftConfiguration
+    );
+
+    // Build a map to dedupe copy operations
+    const fileOperationMap: Map<string, Map<string, boolean>> = new Map<string, Map<string, boolean>>();
+    for (const copyFilesEventAction of eventActions.copyGlobs.get(heftEvent) || []) {
+      for (const globPattern of copyFilesEventAction.globsToCopy) {
+        const resolvedSourceFilePaths: string[] = await this._resolvePathAsync(
+          globPattern,
+          heftConfiguration.buildFolder
+        );
+        for (const resolvedSourceFilePath of resolvedSourceFilePaths) {
+          let resolvedTargetPathsMap: Map<string, boolean> | undefined = fileOperationMap.get(
+            resolvedSourceFilePath
+          );
+          if (!resolvedTargetPathsMap) {
+            resolvedTargetPathsMap = new Map<string, boolean>();
+            fileOperationMap.set(resolvedSourceFilePath, resolvedTargetPathsMap);
+          }
+
+          for (const targetFolder of copyFilesEventAction.targetFolders) {
+            resolvedTargetPathsMap.set(
+              path.resolve(
+                heftConfiguration.buildFolder,
+                targetFolder,
+                path.basename(resolvedSourceFilePath)
```

Comment:
Calling `path.basename()` here seems incorrect. For example, suppose that `globsToCopy` matches two file paths:

- `src/fruit/apple.md` 
- `src/phones/apple.md`

The current implementation will map them both to the path `dist/apple.md`, either silently overwriting the second one, or maybe reporting an error that the other promise is already holding a file lock.

I think most people would expect the output to be:

- `dist/fruit/apple.md` 
- `dist/phones/apple.md`

If there really is a situation where someone would want to condense a nested tree into a flat list (?), that should be an opt-in behavior.  Maybe it should even have its own action kind (`copyAndFlatten`?).

Reply:
The issue with this design is that we would have to be able to specify what folder the path would be relative from, since the copy applies across the project, not _just_ in the `src` folder, so your example would actually produce:
- `dist/src/fruit/apple.md`
- `dist/src/phones/apple.md`

Which is why I went with flatten by default. However, if we knew what path we would want the files relative to (such as the case when the file match glob is `some/path/**/*.md`, then another property would definitely make sense for that. I'm just not sure if that wouldn't be better served by making multiple explicit action definitions, which specify specific directories, ex.
```json
{
      "actionKind": "copyGlobs",
      "heftEvent": "pre-compile",
      "actionId": "defaultCopyGlobs",
      "globsToCopy": ["src/phones/**/*.md"],
      "targetFolders": ["dist/phones"],
      "hardlink": false
},
{
      "actionKind": "copyGlobs",
      "heftEvent": "pre-compile",
      "actionId": "defaultCopyGlobs",
      "globsToCopy": ["src/fruits/**/*.md"],
      "targetFolders": ["dist/fruits"],
      "hardlink": false
}
```

Though... this in and of itself is pretty verbose. It seems like containing it all within one definition would be better, ie.
```json
{
      "actionKind": "copyGlobs",
      "heftEvent": "pre-compile",
      "actionId": "defaultCopyGlobs",
      "copyOperations": [
          {
             "sourceGlobs": [ "src/phones/**/*.md" ],
             "targetFolders": [ "dist/phones" ],
             "hardlink": false
          },
          {
             "sourceGlobs": [ "src/fruits/**/*.md" ],
             "targetFolders": [ "dist/fruits" ],
             "hardlink": false
          }
      ]
}
```

Thoughts?

Reply:
Combining this with some thoughts from @iclanton, what if we did something like this?
```JSON
{
      "actionKind": "copyFiles",
      "heftEvent": "pre-compile",
      "actionId": "defaultCopyFiles",
      "copyOperations": [
          {
             "sourceFolder": "src" // Optional, defaults to project root
             "includeGlobs": [ "**/*.md" ], // Required
             "excludeGlobs": [ "**/__test__/**/*.md" ], // Optional, defaults to []
             "targetFolders": [ "dist" ], // Required
             "hardlink": false, // Optional, defaults to false
             "flatten": false // Optional, defaults to false (true?)
          }
      ]
}
```
This solution has the ability to narrow down the scope of the glob being used, to reduce the perf impact of the search (as per your first comment). It also allows for optionally flattening/retaining the relative path from the `sourceFolder` to the matched file. Additionally, it provides flexibility to specify glob patterns to exclude from included results.

Reply:
I like this design. We can also optimize around some glob patterns like `**/*` and `**/*.<ext>`

Reply:
All these examples still have `**` in them. That is a brute force `O(n)` scan.  The way the plugin is implemented, this scan will be performed repeatedly for each `copyFiles` action. 

Planning to one day cache the scans is a good idea, but let's think about a cache for a moment:

Design 1:
```
  "staticAssetsToCopy": {
    "fileExtensions": [".css", ".jpg", ".png"]
  }
```

Design 2:
```
  "staticAssetsToCopy": {
    "includeGlobs": [
      "src/**/*.css",
      "src/**/*.jpg",
      "src/**/*.png"
    ]
  }
```

Design 1 could easily be optimized as a single scan of the `src` folder that collects files based on their file extension.

Whereas Design 2 requires invoking a complex glob engine multiple times.  The engine has to support negative lookback patterns. It has to be able to scan paths that are not under the `src` folder.  Smart engineers can write their globs efficiently, but the engine needs to support very inefficient globs.

So planning to cache later does not get us out of thinking about performance when designing the input file.